File: server/mcp/feature_extractor.py
"""Feature extractor — local pixel-based hair/skin/eye classification from photos.

Ported from featureExtractor.js. Analyzes uploaded reference photos using
pixel sampling (no AI) to extract hair color, skin tone, and eye color.
Used as fallback when AI vision analysis is unavailable or blocked.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

from PIL import Image

logger = logging.getLogger(__name__)


@dataclass
class ExtractedFeatures:
    hair_color: str | None = None
    skin_tone: str | None = None
    eye_color: str | None = None
    description: str = ""


def _rgb_to_hsl(r: int, g: int, b: int) -> tuple[float, float, float]:
    r_f, g_f, b_f = r / 255, g / 255, b / 255
    mx = max(r_f, g_f, b_f)
    mn = min(r_f, g_f, b_f)
    lum = (mx + mn) / 2

    if mx == mn:
        return (0.0, 0.0, lum * 100)

    d = mx - mn
    s = d / (2 - mx - mn) if lum > 0.5 else d / (mx + mn)

    if mx == r_f:
        h = ((g_f - b_f) / d + (6 if g_f < b_f else 0)) / 6
    elif mx == g_f:
        h = ((b_f - r_f) / d + 2) / 6
    else:
        h = ((r_f - g_f) / d + 4) / 6

    return (h * 360, s * 100, lum * 100)


def _classify_hair_color(h: float, s: float, lum: float) -> str:
    if lum < 15:
        return "very dark black"
    if lum < 25 and s < 30:
        return "black"
    if lum < 30 and 15 < h < 40:
        return "very dark brown"
    if lum < 40 and 10 < h < 45 and s > 20:
        return "dark brown"
    if lum < 50 and 10 < h < 45:
        return "brown"
    if lum < 50 and 5 < h < 20 and s < 25:
        return "dark ash brown"
    if lum < 55 and 15 < h < 40 and s > 30:
        return "medium brown"
    if 40 <= lum < 55 and 20 < h < 50 and s > 40:
        return "auburn"
    if 35 <= lum < 55 and 5 < h < 25 and s > 50:
        return "reddish brown"
    if 40 <= lum < 55 and 10 < h < 30 and s < 25:
        return "ash blonde"
    if 50 <= lum < 70 and 15 < h < 45 and s > 25:
        return "light brown or dark blonde"
    if 55 <= lum < 75 and 20 < h < 50:
        return "dirty blonde"
    if 60 <= lum < 80 and s < 25:
        return "blonde"
    if 60 <= lum < 80 and 20 < h < 50 and s > 30:
        return "golden blonde"
    if 70 <= lum < 90 and s < 20:
        return "light blonde"
    if lum >= 75:
        return "very light blonde or white"
    if 0 < h < 15 and s > 40:
        return "red"
    if 5 < h < 30 and s > 30 and lum < 50:
        return "ginger"
    if s < 15 and lum > 40:
        return "grey or silver"
    return "brown"


def _classify_skin_tone(h: float, s: float, lum: float) -> str:
    if lum < 30:
        return "deep dark brown"
    if lum < 40 and 15 < h < 35:
        return "dark brown"
    if lum < 50 and 15 < h < 40:
        return "warm medium brown"
    if lum < 55 and 10 < h < 35:
        return "medium olive or tan"
    if lum < 60 and 15 < h < 40 and s > 30:
        return "warm light brown"
    if lum < 65 and 10 < h < 35:
        return "light tan or olive"
    if lum < 70 and 15 < h < 40:
        return "warm fair with golden undertones"
    if lum < 75 and s > 20:
        return "fair with warm undertones"
    if lum < 75:
        return "fair with cool undertones"
    if lum < 80:
        return "very fair or porcelain"
    return "very fair or pale"


def _classify_eye_color(r: int, g: int, b: int) -> str:
    h, s, lum = _rgb_to_hsl(r, g, b)
    if lum < 20:
        return "very dark brown, almost black"
    if lum < 30 and s < 40:
        return "dark brown"
    if lum < 40 and s < 50:
        return "brown"
    if lum < 45 and 20 < h < 50 and s > 30:
        return "warm brown with amber flecks"
    if 25 < h < 50 and s > 40 and 35 < lum < 55:
        return "hazel"
    if 20 < h < 45 and s > 50 and lum > 40:
        return "amber"
    if 50 < h < 170 and s > 20 and 25 < lum < 55:
        return "green"
    if 170 < h < 260 and s > 15 and 20 < lum < 50:
        return "blue"
    if 170 < h < 260 and s > 15 and lum >= 50:
        return "light blue"
    if 170 < h < 260 and s < 15:
        return "grey-blue"
    if s < 15 and lum < 40:
        return "dark grey"
    if s < 20 and lum >= 40:
        return "grey"
    return "brown"


def _sample_region(
    pixels: list[tuple[int, int, int, int]],
    img_w: int,
    cx: int,
    cy: int,
    radius: int,
) -> tuple[int, int, int] | None:
    r_sum = g_sum = b_sum = 0
    count = 0
    for dy in range(-radius, radius + 1):
        for dx in range(-radius, radius + 1):
            if dx * dx + dy * dy > radius * radius:
                continue
            x = cx + dx
            y = cy + dy
            if x < 0 or x >= img_w or y < 0 or y >= len(pixels) // img_w:
                continue
            idx = y * img_w + x
            if idx >= len(pixels):
                continue
            a = pixels[idx][3]
            if a < 128:
                continue
            r_sum += pixels[idx][0]
            g_sum += pixels[idx][1]
            b_sum += pixels[idx][2]
            count += 1
    if count == 0:
        return None
    return (round(r_sum / count), round(g_sum / count), round(b_sum / count))


def extract_features(image: Image.Image) -> ExtractedFeatures:
    img = image.convert("RGBA")
    max_dim = 512
    scale = min(max_dim / img.width, max_dim / img.height, 1.0)
    if scale < 1.0:
        new_w = round(img.width * scale)
        new_h = round(img.height * scale)
        img = img.resize((new_w, new_h), Image.Resampling.LANCZOS)

    w, h = img.size
    pixels = img.get_flattened_data()

    features = ExtractedFeatures()

    head_top_y = round(h * 0.05)
    head_cy = round(h * 0.18)
    head_cx = round(w * 0.5)
    head_r = round(min(w, h) * 0.06)

    hair_sample = _sample_region(pixels, w, head_cx, head_top_y, head_r)
    if hair_sample:
        hsl = _rgb_to_hsl(*hair_sample)
        features.hair_color = _classify_hair_color(*hsl)

    forehead_r = round(head_r * 0.5)
    forehead_y = round(head_cy - head_r * 0.3)
    forehead_sample = _sample_region(pixels, w, head_cx, forehead_y, forehead_r)
    if forehead_sample:
        hsl = _rgb_to_hsl(*forehead_sample)
        features.skin_tone = _classify_skin_tone(*hsl)

    eye_y = round(head_cy + head_r * 0.15)
    left_eye_x = round(head_cx - head_r * 0.5)
    right_eye_x = round(head_cx + head_r * 0.5)
    eye_r = max(round(head_r * 0.2), 1)
    left_eye = _sample_region(pixels, w, left_eye_x, eye_y, eye_r)
    right_eye = _sample_region(pixels, w, right_eye_x, eye_y, eye_r)
    eye_sample = left_eye or right_eye
    if eye_sample:
        features.eye_color = _classify_eye_color(*eye_sample)

    parts = []
    if features.hair_color:
        parts.append(f"{features.hair_color} hair")
    if features.skin_tone:
        parts.append(f"{features.skin_tone} skin")
    if features.eye_color:
        parts.append(f"{features.eye_color} eyes")
    features.description = ", ".join(parts)

    return features


def build_photo_derived_design(features: ExtractedFeatures, base_design: str = "") -> str:
    if not features.description:
        return base_design or ""

    design = base_design or "a child character"

    if features.hair_color:
        design = f"{features.hair_color} hair, {design}"

    if features.skin_tone:
        design = f"{features.skin_tone} skin, {design}"

    if features.eye_color:
        design = f"{features.eye_color} eyes, {design}"

    return f"Based on reference photos: {features.description}. {design}"

File: src/lib/featureExtractor.js
// Feature Extractor — analyzes uploaded reference photos locally (canvas-based)
// to extract hair color, skin tone, eye color, and face shape.
// Used when Azure's safety system blocks photo uploads in /images/edits.
// Falls back to rich text descriptions instead of raw image passing.

function loadImage(url) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.crossOrigin = "anonymous";
    img.onload = () => resolve(img);
    img.onerror = () => reject(new Error("Failed to load image"));
    img.src = url;
  });
}

function rgbToHsl(r, g, b) {
  r /= 255; g /= 255; b /= 255;
  const max = Math.max(r, g, b), min = Math.min(r, g, b);
  let h, s, l = (max + min) / 2;

  if (max === min) {
    h = s = 0;
  } else {
    const d = max - min;
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    switch (max) {
      case r: h = ((g - b) / d + (g < b ? 6 : 0)) / 6; break;
      case g: h = ((b - r) / d + 2) / 6; break;
      case b: h = ((r - g) / d + 4) / 6; break;
    }
  }
  return { h: h * 360, s: s * 100, l: l * 100 };
}

function classifyHairColor(h, s, l) {
  if (l < 15) return "very dark black";
  if (l < 25 && s < 30) return "black";
  if (l < 30 && h > 15 && h < 40) return "very dark brown";
  if (l < 40 && h > 10 && h < 45 && s > 20) return "dark brown";
  if (l < 50 && h > 10 && h < 45) return "brown";
  if (l < 50 && h > 5 && h < 20 && s < 25) return "dark ash brown";
  if (l < 55 && h > 15 && h < 40 && s > 30) return "medium brown";
  if (l >= 40 && l < 55 && h > 20 && h < 50 && s > 40) return "auburn";
  if (l >= 35 && l < 55 && h > 5 && h < 25 && s > 50) return "reddish brown";
  if (l >= 40 && l < 60 && h > 10 && h < 30 && s < 25) return "ash blonde";
  if (l >= 50 && l < 70 && h > 15 && h < 45 && s > 25) return "light brown or dark blonde";
  if (l >= 55 && l < 75 && h > 20 && h < 50) return "dirty blonde";
  if (l >= 60 && l < 80 && s < 25) return "blonde";
  if (l >= 60 && l < 80 && h > 20 && h < 50 && s > 30) return "golden blonde";
  if (l >= 70 && l < 90 && s < 20) return "light blonde";
  if (l >= 75) return "very light blonde or white";
  if (h > 0 && h < 15 && s > 40) return "red";
  if (h > 5 && h < 30 && s > 30 && l < 50) return "ginger";
  if (s < 15 && l > 40) return "grey or silver";
  return "brown";
}

function classifySkinTone(h, s, l) {
  if (l < 30) return "deep dark brown";
  if (l < 40 && h > 15 && h < 35) return "dark brown";
  if (l < 50 && h > 15 && h < 40) return "warm medium brown";
  if (l < 55 && h > 10 && h < 35) return "medium olive or tan";
  if (l < 60 && h > 15 && h < 40 && s > 30) return "warm light brown";
  if (l < 65 && h > 10 && h < 35) return "light tan or olive";
  if (l < 70 && h > 15 && h < 40) return "warm fair with golden undertones";
  if (l < 75 && s > 20) return "fair with warm undertones";
  if (l < 75) return "fair with cool undertones";
  if (l < 80) return "very fair or porcelain";
  return "very fair or pale";
}

function classifyEyeColor(r, g, b) {
  const { h, s, l } = rgbToHsl(r, g, b);
  if (l < 20) return "very dark brown, almost black";
  if (l < 30 && s < 40) return "dark brown";
  if (l < 40 && s < 50) return "brown";
  if (l < 45 && h > 20 && h < 50 && s > 30) return "warm brown with amber flecks";
  if (h > 25 && h < 50 && s > 40 && l > 35 && l < 55) return "hazel";
  if (h > 20 && h < 45 && s > 50 && l > 40) return "amber";
  if (h > 50 && h < 170 && s > 20 && l > 25 && l < 55) return "green";
  if (h > 170 && h < 260 && s > 15 && l > 20 && l < 50) return "blue";
  if (h > 170 && h < 260 && s > 15 && l >= 50) return "light blue";
  if (h > 170 && h < 260 && s < 15) return "grey-blue";
  if (s < 15 && l < 40) return "dark grey";
  if (s < 20 && l >= 40) return "grey";
  return "brown";
}

function sampleRegion(imageData, imgW, cx, cy, radius) {
  const data = imageData.data;
  const samples = [];
  for (let dy = -radius; dy <= radius; dy++) {
    for (let dx = -radius; dx <= radius; dx++) {
      if (dx * dx + dy * dy > radius * radius) continue;
      const x = Math.round(cx + dx);
      const y = Math.round(cy + dy);
      if (x < 0 || x >= imgW || y < 0 || y >= imageData.height) continue;
      const idx = (y * imgW + x) * 4;
      const a = data[idx + 3];
      if (a < 128) continue;
      samples.push({ r: data[idx], g: data[idx + 1], b: data[idx + 2] });
    }
  }
  if (samples.length === 0) return null;
  const avg = samples.reduce((acc, c) => ({ r: acc.r + c.r, g: acc.g + c.g, b: acc.b + c.b }), { r: 0, g: 0, b: 0 });
  return { r: Math.round(avg.r / samples.length), g: Math.round(avg.g / samples.length), b: Math.round(avg.b / samples.length) };
}

export async function extractFeaturesFromPhotos(photoUrls) {
  if (!photoUrls || photoUrls.length === 0) return null;

  const features = {
    hair_color: null,
    skin_tone: null,
    eye_color: null,
    description: "",
  };

  try {
    const img = await loadImage(photoUrls[0]);
    const canvas = document.createElement("canvas");
    const maxDim = 512;
    const scale = Math.min(maxDim / img.naturalWidth, maxDim / img.naturalHeight, 1);
    canvas.width = Math.round(img.naturalWidth * scale);
    canvas.height = Math.round(img.naturalHeight * scale);
    const ctx = canvas.getContext("2d");
    ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    const w = canvas.width;
    const h = canvas.height;

    const headTopy = Math.round(h * 0.05);
    const headCy = Math.round(h * 0.18);
    const headCx = Math.round(w * 0.5);
    const headR = Math.round(Math.min(w, h) * 0.06);

    // Hair: sample from top of head
    const hairSample = sampleRegion(imageData, w, headCx, headTopy, headR);
    if (hairSample) {
      const hsl = rgbToHsl(hairSample.r, hairSample.g, hairSample.b);
      features.hair_color = classifyHairColor(hsl.h, hsl.s, hsl.l);
    }

    // Skin: sample from forehead area
    const foreheadSample = sampleRegion(imageData, w, headCx, headCy - headR * 0.3, Math.round(headR * 0.5));
    if (foreheadSample) {
      const hsl = rgbToHsl(foreheadSample.r, foreheadSample.g, foreheadSample.b);
      features.skin_tone = classifySkinTone(hsl.h, hsl.s, hsl.l);
    }

    // Eyes: sample from expected eye region
    const eyeY = Math.round(headCy + headR * 0.15);
    const leftEyeX = Math.round(headCx - headR * 0.5);
    const rightEyeX = Math.round(headCx + headR * 0.5);
    const eyeR = Math.round(headR * 0.2);
    const leftEye = sampleRegion(imageData, w, leftEyeX, eyeY, eyeR);
    const rightEye = sampleRegion(imageData, w, rightEyeX, eyeY, eyeR);
    const eyeColor = leftEye || rightEye;
    if (eyeColor) {
      features.eye_color = classifyEyeColor(eyeColor.r, eyeColor.g, eyeColor.b);
    }

    // Build description
    const parts = [];
    if (features.hair_color) parts.push(`${features.hair_color} hair`);
    if (features.skin_tone) parts.push(`${features.skin_tone} skin`);
    if (features.eye_color) parts.push(`${features.eye_color} eyes`);
    features.description = parts.join(", ");

    return features;
  } catch (err) {
    console.error("Feature extraction failed:", err);
    return null;
  }
}

export function buildPhotoDerivedDesign(features, baseDesign) {
  if (!features || !features.description) return baseDesign || "";

  let design = baseDesign || "a child character";

  if (features.hair_color) {
    design = design.replace(/\b(blonde|brown|black|red|ginger|auburn|white|grey|gray|dark brown|light brown|dirty blonde|strawberry blonde|wavy|curly|straight|shoulder.length|long|short)\b[^,]*(hair)/gi, `${features.hair_color} hair`);
    if (!/hair/i.test(design)) {
      design = `${features.hair_color} hair, ${design}`;
    }
  }

  if (features.skin_tone) {
    design = design.replace(/\b(fair|pale|light|medium|tan|olive|brown|dark|deep)\s*(skin|complexion|tone)\b[^,]*/gi, `${features.skin_tone} skin`);
    design = design.replace(/\b(skin tone|complexion)\s*:\s*[^,)]+/gi, `skin tone: ${features.skin_tone}`);
    if (!/skin/i.test(design)) {
      design = `${features.skin_tone} skin, ${design}`;
    }
  }

  if (features.eye_color) {
    design = design.replace(/\b(hazel|brown|blue|green|grey|gray|amber|dark)\s*(eyes|eye color)\b[^,]*/gi, `${features.eye_color}`);
    design = design.replace(/\b(eye color|eyes)\s*:\s*[^,)]+/gi, `eyes: ${features.eye_color}`);
    if (!/eye/i.test(design)) {
      design = `${features.eye_color} eyes, ${design}`;
    }
  }

  return `Based on reference photos: ${features.description}. ${design}`;
}
